File: src/utils/adminPassword.ts
// Typed wrappers around the Rust-side admin-password IPC commands.
//
// The actual secret lives in the OS keychain (Secret Service / macOS
// Keychain / Windows Credential Manager) via the `keyring` crate on the
// Rust side. The JS layer only ever sees:
//
//   - a boolean status ("is something stored?")  via admin_password_status
//   - a setter that takes the password and returns nothing
//   - an idempotent clear
//
// The password value never crosses the IPC boundary from Rust back to
// JS. The Tauri commands are wired in src-tauri/src/commands/mod.rs and
// registered in lib.rs.

import { invoke } from "@tauri-apps/api/core";

/**
 * Sentinel string the Rust executor returns from `execute_command` when
 * the user requested an elevated run on Unix but no password is stored
 * in the keychain yet. The JS bridge compares with `===` (strict
 * equality) — any wrapping like `new Error(ADMIN_PASSWORD_REQUIRED)`
 * would still expose `.message` matching this exact value.
 *
 * Mirrored as `ERR_ADMIN_PASSWORD_REQUIRED` in Rust; both sides have a
 * unit test pinning the literal so a rename on one side breaks the
 * other side's tests, not production silently.
 */
export const ADMIN_PASSWORD_REQUIRED = "ADMIN_PASSWORD_REQUIRED";

/**
 * Prefix the Rust executor returns when reading the keychain failed
 * (Linux without D-Bus, macOS denied access, etc.). The remainder of
 * the message is suitable for surfacing to the user as context — the
 * keychain library does NOT include the password value in error
 * variants.
 */
export const ADMIN_PASSWORD_BACKEND_PREFIX = "ADMIN_PASSWORD_BACKEND:";

/**
 * @returns true iff a sudo password is currently stored in the OS
 *   keychain for this app. Used by the Settings UI to swap the
 *   "Set password" / "Clear saved password" buttons, and by the
 *   CommandForm to decide whether to show a hint about the first
 *   elevated run.
 */
export async function hasAdminPassword(): Promise<boolean> {
  return invoke<boolean>("admin_password_status");
}

/**
 * Persist the given password to the OS keychain. Whitespace is trimmed
 * Rust-side; empty strings are rejected with a typed error. The Promise
 * resolves to `void` — there is no read-back of the stored value.
 *
 * Throws a string error from the Rust handler on failure (keychain
 * unavailable, OS denied, etc.).
 */
export async function setAdminPassword(password: string): Promise<void> {
  await invoke("set_admin_password", { password });
}

/**
 * Remove the stored password. Idempotent — calling it when nothing is
 * stored is not an error, so the UI can call it unconditionally.
 */
export async function clearAdminPassword(): Promise<void> {
  await invoke("clear_admin_password");
}

/**
 * True when the given error matches the `ADMIN_PASSWORD_REQUIRED`
 * sentinel from Rust. Centralises the comparison so callers don't
 * have to reach into `err.message` themselves and so a future move to
 * a typed AppError variant only touches this one spot.
 */
export function isAdminPasswordRequiredError(err: unknown): boolean {
  const msg =
    err instanceof Error
      ? err.message
      : typeof err === "string"
        ? err
        : String(err);
  return msg === ADMIN_PASSWORD_REQUIRED;
}

File: src-tauri/src/security/admin_password.rs
// Storage helpers for the sudo password used by the "Run as administrator"
// feature on Unix.
//
// The password lives in the OS-provided keychain — Secret Service on
// Linux, the macOS Keychain, or the Windows Credential Manager — via
// the `keyring` crate. We never persist it in SQLite, in JSON config,
// or in any Tauri-emitted event, and we never log either the value or
// its length.
//
// All entries share a fixed (service, username) pair:
//   service:  "app.procmix.desktop"
//   account:  "admin-sudo"
//
// Tests run against `keyring`'s in-memory mock backend so CI never
// touches a real keychain. The default backend is restored after each
// test via a guard so tests cannot leak state into each other.

use keyring::Entry;
use thiserror::Error;

/// Identifier used when registering and looking up the credential.
/// Kept stable across releases — changing it would orphan stored
/// passwords on the user's machine.
const SERVICE: &str = "app.procmix.desktop";
const ACCOUNT: &str = "admin-sudo";

/// Errors surfaced by this module.
///
/// `NoEntry` is intentionally NOT modelled as an error: callers want
/// `Option<String>` semantics for "is there a password stored?".
#[derive(Debug, Error)]
pub enum AdminPasswordError {
    /// Underlying keychain backend failure (service unavailable, no
    /// session bus on Linux headless, permission denied, etc.). The
    /// inner message is suitable for logging but NOT for showing the
    /// password value itself — `keyring` never includes the secret in
    /// its error variants.
    #[error("keychain backend error: {0}")]
    Backend(String),
}

fn entry() -> Result<Entry, AdminPasswordError> {
    Entry::new(SERVICE, ACCOUNT).map_err(|e| AdminPasswordError::Backend(e.to_string()))
}

/// Read the currently-stored admin password.
///
/// Returns `Ok(None)` when the keychain reports no entry — this is
/// the normal state on first launch and is NOT an error. Any other
/// failure (D-Bus unavailable, OS denied access, malformed entry)
/// surfaces as [`AdminPasswordError::Backend`].
pub fn get() -> Result<Option<String>, AdminPasswordError> {
    let entry = entry()?;
    match entry.get_password() {
        Ok(s) => Ok(Some(s)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(AdminPasswordError::Backend(e.to_string())),
    }
}

/// Persist the given password. Empty strings are rejected — sudo
/// would never accept one anyway, and silently storing one would
/// hide bugs in the UI.
pub fn set(password: &str) -> Result<(), AdminPasswordError> {
    if password.is_empty() {
        return Err(AdminPasswordError::Backend(
            "password cannot be empty".to_string(),
        ));
    }
    let entry = entry()?;
    entry
        .set_password(password)
        .map_err(|e| AdminPasswordError::Backend(e.to_string()))
}

/// Remove the stored password if any. A missing entry is not an error
/// — `clear()` is idempotent so the UI can call it unconditionally.
pub fn clear() -> Result<(), AdminPasswordError> {
    let entry = entry()?;
    match entry.delete_credential() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(AdminPasswordError::Backend(e.to_string())),
    }
}

/// Convenience wrapper: `true` iff a password is currently stored.
/// The boolean shape matches the JS-side `hasAdminPassword()` API,
/// avoiding the need to leak the password value across IPC just to
/// answer "is something there?".
pub fn has() -> Result<bool, AdminPasswordError> {
    Ok(get()?.is_some())
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `set("")` must reject before touching the keychain. This guards
    /// the boundary check against accidental removal in a refactor —
    /// without it, sudo would later reject an empty password noisily
    /// and the user would have no idea what they typed.
    ///
    /// Note: we don't unit-test the keychain round-trip itself. The
    /// `keyring` crate's mock backend deliberately isolates each
    /// `Entry::new` call to a fresh in-memory store, so a `set` on one
    /// entry is invisible to a `get` on another — exactly the opposite
    /// of how real platform backends behave. Testing against the mock
    /// would lock in semantics we don't ship. Real-backend coverage
    /// happens at the manual QA / smoke-test layer.
    #[test]
    fn set_rejects_empty_password_without_touching_keychain() {
        let err = set("").unwrap_err();
        match err {
            AdminPasswordError::Backend(msg) => {
                assert!(
                    msg.contains("empty"),
                    "expected 'empty' in message, got: {msg}"
                );
            }
        }
    }
}
